### PR TITLE
Replace the check for Executable using the new variable HasRuntimeOutput and _IsExecutable

### PR DESCRIPTION
**Pull request by basoundr**
This variable is now used in places where we check to see if the project is an executable to
enable Auto Binding Redirect generation and F5 run.

This is in response to #1174 

Tagging @nguerrera @srivatsn @davkean @dotnet/project-system for review

### CODE REVIEW

**Comment by nguerrera** [comment]
@eerhardt @terrajobst This aligns with the discussion on https://github.com/dotnet/sdk/issues/1070

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 1]
```diff
@@ -20,18 +20,19 @@ Copyright (c) .NET Foundation. All rights reserved.
...
           Should be re-visited when multiple TFM support is added to Dependencies logic.
     -->
     <DefaultImplicitPackages>Microsoft.NETCore.App;NETStandard.Library</DefaultImplicitPackages>
   </PropertyGroup>
 
-
   <UsingTask TaskName="GetNearestTargetFramework" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
   <UsingTask TaskName="NETSdkError" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
-  
+
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: nit: unnecessary whitespace diff

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 2]
```diff
@@ -20,18 +20,19 @@ Copyright (c) .NET Foundation. All rights reserved.
     <MicrosoftNETBuildTasksTFM Condition=" '$(MicrosoftNETBuildTasksTFM)' == ''">net46</MicrosoftNETBuildTasksTFM>
     <MicrosoftNETBuildTasksDirectory>$(MicrosoftNETBuildTasksDirectoryRoot)$(MicrosoftNETBuildTasksTFM)/</MicrosoftNETBuildTasksDirectory>
     <MicrosoftNETBuildTasksAssembly>$(MicrosoftNETBuildTasksDirectory)Microsoft.NET.Build.Tasks.dll</MicrosoftNETBuildTasksAssembly>
-    
-    <!-- 
+    <IsExecutable Condition="'$(IsExecutable)' == '' and '$(OutputType)'=='Exe'">true</IsExecutable>
+    <IsExecutable Condition="'$(IsExecutable)' == '' and '$(OutputType)'=='WinExe'">true</IsExecutable>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: I believe this is after the use in BeforeCommon.targets and so the change won't have the effect you want. 

Mind you, I'm surprised this test didn't fail: https://github.com/dotnet/sdk/blob/1d81050cab00c6f9ab8bf5b984bc42eff2760c1e/test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs#L351 

I'm surprised to see it doesn't check the contents of the app.config. Can you check  that before and after your change?

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Nevermind, it did fail. :)

**Review by nguerrera — CHANGES_REQUESTED** [review]
See above.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets (hunk 2 above)
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: We should move this to BeforeCommon.targets to fix it.

**Comment by nguerrera** [comment]
LGTM in principle once the ordering issue is fixed and tests pass.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 3]
```diff
@@ -304,7 +304,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       </PropertyGroup>
     </When>
     
-    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(OutputType)' == 'Exe'">
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(IsExecutable)' == 'true'">
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: I was under the assumption that things like Test projects were going to set `$(IsExecutable)` to true.  If that's the case, then I don't think these two lines should be changed.

The "Run Information" is used during F5 and `dotnet run`.  If you try to `dotnet run` a Test project, will it try to execute the Test's $(TargetPath), which is a .dll on .NET Framework.  Thus failing with a bad error.

Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Also, if Test projects are expected to set this property, I think we should think about the best name to use for this property.

**Comment by srivatsn** [comment]
@codito @sbaid - This PR is introducing a IsExecutable property to denote a project as executable. With this, I think you can set IsExecutable=true in your project and set OutputType=dll. That way, I think you don't have to do the Main method generation but still have deps file be generated. Would that work?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets (hunk 3 above)
Comment (tannergooding) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: @eerhardt, I would think this would be valid and part of setting `IsExecutable=true` would involve having uses configure the F5 information (to use xunit, nunit, dotnet, etc...)

However, maybe having a separate variable to disable (or explicitly enable) F5 for some scenarios would be desirable

Comment (basoundr) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: @eerhardt 

> I was under the assumption that things like Test projects were going to set $(IsExecutable) to true. 

Yes, thats true.

>  If that's the case, then I don't think these two lines should be changed.

We still need it so the Windows Application output type projects can still be run

> I think we should think about the best name to use for this property.

I am open for a better name. But I feel `IsExecutable` conveys its purpose to the point.

Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Test projects aren't "executable".  You can't double-click on it's output and expect the test to run.  You can't type into a command-line `C:\mytest\mytest.dll` and watch the tests run.

> I am open for a better name.

The list I came up with before (which we can add more to) was:

- TreatAsExecutable
- HasRuntimeOutput
- HasRunnableOutput
- HasExecutableOutput
- IsExecutable
- IsRunnable

I think the one I like most, given all the situations this is going to be used, is what we landed on in the project.json world: `HasRuntimeOutput`.  The output isn't "runnable" or "executable", but instead the output is associated with a runtime, and that's why we can generate binding redirects and deps.json files that will work at runtime.

> I would think this would be valid and part of setting IsExecutable=true would involve having uses configure the F5 information (to use xunit, nunit, dotnet, etc...)

That's exactly my point.  Look at the rest of the code around this area that I commented on.  If `IsExecutable=true` on a test project, the "Run Information" being returned says that you can just run the output directly:

```XML
<PropertyGroup>
  <RunCommand Condition="'$(RunCommand)' == ''">$(TargetPath)</RunCommand>
  <RunArguments Condition="'$(RunArguments)' == ''">$(StartArguments)</RunArguments>
```

> However, maybe having a separate variable to disable (or explicitly enable) F5 for some scenarios would be desirable

Again, that's what the `RunCommand` property is for.  If it is set, you can `dotnet run` or F5 this project.  If it isn't set, then you can't.

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Thanks, Eric. Here's what I propose based on your feedback.

_IsExecutable : "private" variable that's true iff output type is exe or winexe. Use this everywhere we have OutputType == Exe. This is not an extensibility point, it is just for our own maintainability so that we don't reintroduce winexe bugs. Controls getting a run command that just launches the output exe.

HasRuntimeOutput: defaults to _IsExecutable, but overridable by test project or similar. Controls binding redirects and other runtime configuration file generation.

If something other than a (Win)Exe wants to be runnable, it supplies a custom RunCommand as it already works.

Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Sorry, I meant to reply to this earlier.

This sounds like a great plan.  Thanks.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets [hunk 4]
```diff
@@ -18,7 +18,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <_GetChildProjectCopyToPublishDirectoryItems Condition="'$(_GetChildProjectCopyToPublishDirectoryItems)' == ''">true</_GetChildProjectCopyToPublishDirectoryItems>
 
     <!-- publishing self-contained apps should publish the native host as $(AssemblyName).exe -->
-    <DeployAppHost Condition=" '$(DeployAppHost)' == '' and '$(OutputType)' == 'Exe' and '$(RuntimeIdentifier)' != '' and '$(SelfContained)' == 'true'">true</DeployAppHost>
+    <DeployAppHost Condition=" '$(DeployAppHost)' == '' and '$(HasRuntimeOutput)' == 'true' and '$(RuntimeIdentifier)' != '' and '$(SelfContained)' == 'true'">true</DeployAppHost>
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: This should stay `'$(OutputType)' == 'Exe'` or $(_IsExecutable)..  Tests won't get a `$Test.exe`.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 5]
```diff
@@ -210,7 +210,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     it requires a host in the output directory to load the app.
     During "publish", all required assets are copied to the publish directory.
     -->
-    <ItemGroup Condition="'$(SelfContained)' == 'true' and '$(RuntimeIdentifier)' != '' and '$(OutputType)' == 'Exe'">
+    <ItemGroup Condition="'$(SelfContained)' == 'true' and '$(RuntimeIdentifier)' != '' and '$(HasRuntimeOutput)' == 'true'">
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Should stay '$(OutputType)' == 'Exe' or $(_IsExecutable).

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 6]
```diff
@@ -255,7 +255,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       </NativeNETCoreCopyLocalItems>
     </ItemGroup>
 
-    <ItemGroup Condition="'$(SelfContained)' == 'true' and '$(RuntimeIdentifier)' != '' and '$(OutputType)' == 'Exe'">
+    <ItemGroup Condition="'$(SelfContained)' == 'true' and '$(RuntimeIdentifier)' != '' and '$(HasRuntimeOutput)' == 'true'">
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Should stay '$(OutputType)' == 'Exe' or $(_IsExecutable).

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 7]
```diff
@@ -288,7 +288,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       </PropertyGroup>
     </When>
     
-    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(IsExecutable)' == 'true'">
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(HasRuntimeOutput)' == 'true'">
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Should stay '$(OutputType)' == 'Exe' or $(_IsExecutable).  I discussed this prevously that tests are not `dotnet run` able, becuase you can't just run the output assembly.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 8]
```diff
@@ -304,7 +304,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       </PropertyGroup>
     </When>
     
-    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(IsExecutable)' == 'true'">
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(HasRuntimeOutput)' == 'true'">
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Should stay '$(OutputType)' == 'Exe' or $(_IsExecutable).

**Review by eerhardt — CHANGES_REQUESTED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets (hunk 8 above)
Comment (basoundr) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Using _IsExecutable since that was original intent of this PR

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets (hunk 7 above)
Comment (basoundr) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Using _IsExecutable since that was original intent of this PR

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets (hunk 5 above)
Comment (basoundr) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Seems _IsExecutable is the one we are looking for since we need to handle Winexe here as well

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets (hunk 6 above)
Comment (basoundr) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Seems _IsExecutable is the one we are looking for since we need to handle Winexe here as well

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets (hunk 5 above)
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: My comment was a little vague.  I meant this should either:

- stay '$(OutputType)' == 'Exe'
or
- use $(_IsExecutable)

I didn't mean the MSBuild code should be ` '$(OutputType)' == 'Exe' or $(_IsExecutable)`

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: I want to use $(_IsExecutable) in all these cases unless there's a reason to exclude WinExe but I don't think there is any such case. I think every case we did not treat exe and winexe the same was a bug.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets [hunk 9]
```diff
@@ -13,6 +13,9 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <PropertyGroup>
     <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+    <_IsExecutable Condition="'$(_IsExecutable)' == '' and '$(OutputType)'=='Exe'">true</_IsExecutable>
+    <_IsExecutable Condition="'$(_IsExecutable)' == '' and '$(OutputType)'=='WinExe'">true</_IsExecutable>
+    <HasRuntimeOutput Condition="'$(HasRuntimeOutput)' == ''">_IsExecutable</HasRuntimeOutput>
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets]: This should be 

`<HasRuntimeOutput Condition="'$(HasRuntimeOutput)' == ''">$(_IsExecutable)</HasRuntimeOutput>`

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 10]
```diff
@@ -33,7 +33,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   </PropertyGroup>
 
   <PropertyGroup>
-    <GenerateRuntimeConfigurationFiles Condition=" '$(GenerateRuntimeConfigurationFiles)' == '' and '$(OutputType)' == 'exe' ">true</GenerateRuntimeConfigurationFiles>
+    <GenerateRuntimeConfigurationFiles Condition=" '$(GenerateRuntimeConfigurationFiles)' == '' and '$(HasRuntimeOutput)' == 'true' ">true</GenerateRuntimeConfigurationFiles>
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: This should actually have a `Is .NET Core` check on it too.  RuntimeConfigurationFiles don't make sense for desktop or mono.

But I don't think you have to make the change here.  Someone should at some point.

Comment (basoundr) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: Opened #1183

**Comment by basoundr** [comment]
@dotnet-bot retest this please

**Review by eerhardt — APPROVED** [review]
Thanks, @basoundr.

**Comment by basoundr** [comment]
@dotnet-bot retest Windows_NT_FullFramework Release please

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets [hunk 11]
```diff
@@ -403,7 +403,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   ============================================================
   -->
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(OutputType)' == 'Exe'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(HasRuntimeOutput)' == 'true'">
     <ProjectCapability Include="CrossPlatformExecutable" />
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets]: This will light up the docker integration in VS, which I don't think should happen for test projects once they switch to not being actual exes. @codito, @saurabsa Thoughts?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets [hunk 12]
```diff
@@ -20,8 +20,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <MicrosoftNETBuildTasksTFM Condition=" '$(MicrosoftNETBuildTasksTFM)' == ''">net46</MicrosoftNETBuildTasksTFM>
     <MicrosoftNETBuildTasksDirectory>$(MicrosoftNETBuildTasksDirectoryRoot)$(MicrosoftNETBuildTasksTFM)/</MicrosoftNETBuildTasksDirectory>
     <MicrosoftNETBuildTasksAssembly>$(MicrosoftNETBuildTasksDirectory)Microsoft.NET.Build.Tasks.dll</MicrosoftNETBuildTasksAssembly>
-    
-    <!-- 
+
+    <!--
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.Common.targets]: Nit: This file has only whitespace diffs. Better to rebase them away to keep the history clean.

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
@dotnet-bot Test Windows_NT_FullFramework Release

**Comment by codito** [comment]
> @codito @sbaid - This PR is introducing a IsExecutable property to denote a project as executable. With this, I think you can set IsExecutable=true in your project and set OutputType=dll. That way, I think you don't have to do the Main method generation but still have deps file be generated. Would that work?

@srivatsn, @basoundr yes, we can remove the entrypoint generation code. I think just setting `HasRuntimeOutput` and `OutputType=dll` may not work, because of [this check](https://github.com/dotnet/sdk/pull/1178/files#diff-ff37ca0d83837d864f8a3079a225a9b7R16); we can set `_IsExecutable` also, but it seems like a private property.

**Comment by basoundr** [comment]
@dotnet-bot Test Windows_NT_FullFramework Release

**Comment by livarcocc** [comment]
@MattGertz for approval.

**Comment by basoundr** [comment]
@codito 
> I think just setting HasRuntimeOutput and OutputType=dll may not work

The condition for deps generation and Binding redirects are based on the value of `HasRuntimeOutput`. So, as long as you set the value of `HasRuntimeOutput` to true, everything should just work.
